File: problems/0003_longest-substring-without-repeating-characters/Solution.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int max = 0;
        for (int i = 0, start = 0, end = 0; i < s.length(); ++i) {
            final int idx = find(s, s.charAt(i), start, end);
            ++end;
            if (idx == -1)
                max = Math.max(max, end - start);
            else
                start = idx + 1;
        }
        return max;
    }
    static int find(final String s, final char c, final int start, final int end) {
        for (int i = start; i < end; ++i)
            if (s.charAt(i) == c)
                return i;
        return -1;
    }
}

File: problems/0003_longest-substring-without-repeating-characters/Solution.rs
use std::cmp;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut max = 0;
        let (mut start, mut end) = (0 as usize, 0 as usize);
        for c in s.chars() {
            let x = Self::find(s.as_ptr(), c, start, end);
            end += 1;
            if let Some(idx) = x {
                start = idx + 1;
            } else {
                max = cmp::max(max, (end - start) as i32);
            }
        }
        max
    }
    #[inline]
    fn find(ptr: *const u8, c: char, start: usize, end: usize) -> Option<usize> {
        for i in start..end {
            if unsafe { *(ptr.add(i)) } as char == c {
                return Some(i);
            }
        }
        None
    }
}
